RecoParticleFlow/PFClusterProducer/plugins/alpaka/PFClusterSoAProducerKernel.dev.cc

```diff
@@ -0,0 +1,1652 @@
+#include <alpaka/alpaka.hpp>
+
+#include "FWCore/Utilities/interface/bit_cast.h"
+#include "HeterogeneousCore/AlpakaInterface/interface/config.h"
+#include "HeterogeneousCore/AlpakaInterface/interface/workdivision.h"
+
+#include "DataFormats/ParticleFlowReco/interface/PFLayer.h"
+#include "RecoParticleFlow/PFClusterProducer/plugins/alpaka/PFClusterSoAProducerKernel.h"
+#include "RecoParticleFlow/PFClusterProducer/plugins/alpaka/PFClusterECLCC.h"
+
+namespace ALPAKA_ACCELERATOR_NAMESPACE {
+
+  using namespace cms::alpakatools;
+
+  using namespace reco::pfClustering;
+
+  static const int threadsPerBlockForClustering = std::is_same_v<Device, alpaka::DevCpu> ? 32 : 512;
+
+  // Calculation of dR2 for Clustering
+  ALPAKA_FN_ACC ALPAKA_FN_INLINE static float dR2(Position4 pos1, Position4 pos2) {
+    float mag1 = sqrtf(pos1.x * pos1.x + pos1.y * pos1.y + pos1.z * pos1.z);
+    float cosTheta1 = mag1 > 0.0 ? pos1.z / mag1 : 1.0;
+    float eta1 = 0.5 * logf((1.0 + cosTheta1) / (1.0 - cosTheta1));
+    float phi1 = atan2f(pos1.y, pos1.x);
+
+    float mag2 = sqrtf(pos2.x * pos2.x + pos2.y * pos2.y + pos2.z * pos2.z);
+    float cosTheta2 = mag2 > 0.0 ? pos2.z / mag2 : 1.0;
+    float eta2 = 0.5 * logf((1.0 + cosTheta2) / (1.0 - cosTheta2));
+    float phi2 = atan2f(pos2.y, pos2.x);
+
+    float deta = eta2 - eta1;
+    float dphi = std::abs(std::abs(phi2 - phi1) - M_PI) - M_PI;
+    return (deta * deta + dphi * dphi);
+  }
+
+#if defined(__CUDA_ARCH__) or defined(__HIP_DEVICE_COMPILE__)
+  template <typename TAcc, typename = std::enable_if_t<alpaka::isAccelerator<TAcc>>>
+  static __device__ __forceinline__ float atomicMaxF(const TAcc& acc, float* address, float val) {
+    int ret = __float_as_int(*address);
+    while (val > __int_as_float(ret)) {
+      int old = ret;
+      if ((ret = atomicCAS((int*)address, old, __float_as_int(val))) == old)
+        break;
+    }
+    return __int_as_float(ret);
+  }
+#else
+  template <typename TAcc, typename = std::enable_if_t<alpaka::isAccelerator<TAcc>>>
+  ALPAKA_FN_ACC ALPAKA_FN_INLINE static float atomicMaxF(const TAcc& acc, float* address, float val) {
+    // CPU implementation uses edm::bit_cast
+    int ret = edm::bit_cast<int>(*address);
+    while (val > edm::bit_cast<float>(ret)) {
+      int old = ret;
+      if ((ret = alpaka::atomicCas(acc, (int*)address, old, edm::bit_cast<int>(val))) == old)
+        break;
+    }
+    return edm::bit_cast<float>(ret);
+  }
+#endif
```

Comment:
could you move this to `HeterogeneousCore/AlpakaInterface/interface/atomicMaxF.h` ?